frontend_vue/src/utils/incidentAlertService.ts

```diff
@@ -0,0 +1,176 @@
+      ]
+    : 'HTTP';
+}
+
+/**
+ * Formats a given string by replacing underscores with spaces and capitalizing the first letter.
+ */
+function formatKey(string: string) {
+  const formattedString = string
+    .replace(/_/g, ' ')
+    .replace(/^(.)/, function (match, group) {
+      return group.toUpperCase();
+    });
+
+  return formattedString;
+}
+
+/**
+ * Formats the incident details object by applying custom labels and formatting keys.
+ */
+export function formatLabels(
+  incidentDetails: FormattedHitsType | keyof FormattedHitsType
+) {
+  try {
+    return Object.entries(incidentDetails).reduce(
+      (acc: Record<string, unknown>, [key, val]) => {
+        const formattedKey = INCIDENT_DETAIL_CUSTOM_LABELS.hasOwnProperty(key)
+          ? INCIDENT_DETAIL_CUSTOM_LABELS[
+              key as keyof typeof INCIDENT_DETAIL_CUSTOM_LABELS
+            ]
+          : formatKey(key);
+
+        if (typeof val === 'object' && val !== null) {
+          acc[formattedKey] = formatLabels(
+            val as unknown as keyof FormattedHitsType
+          );
+        } else {
+          acc[formattedKey] = val;
+        }
+
+        return acc;
+      },
+      {}
+    );
+  } catch (error) {
+    console.error(`Error in formatting labels: ${error}`);
+    return incidentDetails;
+  }
+}
+
+/**
+ * Formats the label of a token type based on the given token.
+ */
+export function formatTokenTypeLabel(token: string) {
+  const token_const: string | undefined = Object.values(TOKENS_TYPE).find(
+    (value) => value === token
+  );
+  return token_const ? tokenServices[token_const].label : token;
+}
+
+/**
+ * Removes keys and values from an object if the value is null, an empty object, an empty array,
+ * or if all nested keys and values are removed.
+ */
+export function removeNullEmptyObjectsAndArrays(
+  incidentDetails: FormattedHitsType
+): FormattedHitsType {
+  try {
+    return Object.entries(incidentDetails).reduce((acc, [key, value]) => {
+      if (
+        value !== null &&
+        value !== undefined &&
+        !(typeof value === 'object' && Object.keys(value).length === 0) &&
+        !(Array.isArray(value) && value.length === 0)
+      ) {
+        if (typeof value === 'object') {
+          const nestedObj = removeNullEmptyObjectsAndArrays(
+            value as FormattedHitsType
+          );
+          if (Object.keys(nestedObj).length > 0) {
+            acc[key] = nestedObj as Record<string, string>;
+          }
+        } else {
+          acc[key] = value;
+        }
+      }
+      return acc;
+    }, {} as FormattedHitsType);
+  } catch (error) {
+    console.error(`Error in remove Null, Empty Objects and Arrays: ${error}`);
+    return incidentDetails;
+  }
+}
+
+function isCreditCardtoken(token: string) {
+  return token === TOKENS_TYPE.CREDIT_CARD;
+}
+
+function isAWStoken(token: string) {
+  return token === TOKENS_TYPE.AWS_KEYS;
+}
+
+export function buildIncidentDetails(
```

Comment:
same for this function, what do we get from doing this here instead of in the rendering step? most of this is just `'thing': hitAlert.thing`?

Reply:
most of them, but not all of them.
We have many conditional fields here, and I'm happy to deal with them before I pass the object to the UI.
I'd have to check those conditions on the template side otherwise. This would add a lot of complexity and would make the template cluttered and hard to follow

ie.
`geo_info: !hitAlert.geo_info.bogon
        ? {
            ...hitAlert.geo_info,
          }
        : {
            ip: hitAlert.geo_info.ip,
            bogon: hitAlert.geo_info.bogon,
          },`

Reply:
we can try it. i in my head i figured it would be clearer in the html

Reply:
I'm really not sure how to handle this inside the template.
It means I'll have to add checks at many points in the nesting loops, checking for specific key/vals

for instance for the bogon, I will have to add something like 
`                <ul
                  v-if="isNotEmpty(nested_val) && hitAlert.geo_info.bogon"
                  class="break-words"
                >
                  <IncidentDetailsListItem
                    :label="nested_key"
                    :value="hitAlert.geo_info.bogon"
                  />
                  <IncidentDetailsListItem
                    :label="nested_key"
                    :value="hitAlert.geo_info.ip,"
                  />
                  </ul>
               <template v-else>
               <IncidentDetailsListItem
                    v-if="!isObject(nested_val) "
                    :label="nested_key"
                    :value="nested_val"
                  />
                  </template>
                  ... move forward
                  
 We also add an extra 'basic info' key, which arbitrarily gathers some values that the backend gives us not ordered
       basic_info: {
        token_type: formatTokenTypeLabel(hitAlert.token_type),
        input_channel: hasChannelCustomLabel(hitAlert.input_channel),
        src_data: hitAlert.src_data,
        useragent: hitAlert.useragent || null,
        etc
      },`
 `

Reply:
There's no way I can think of, that this would look fine on a template

Reply:
let's skip it. by "we can try it" i meant lets go with your code heh